plugins/common/adx/adx.go

```diff
@@ -0,0 +1,206 @@
+	metricIngestor, err := adx.GetMetricIngestor(ctx, tableName)
+	if err != nil {
+		return err
+	}
+
+	reader := bytes.NewReader(metricsArray)
+	mapping := ingest.IngestionMappingRef(tableName+"_mapping", ingest.JSON)
+	if metricIngestor != nil {
+		if _, err := metricIngestor.FromReader(ctx, reader, format, mapping); err != nil {
+			return fmt.Errorf("sending ingestion request to Azure Data Explorer for table %q failed: %w", tableName, err)
+		}
+	}
+	return nil
+}
+
+func (adx *Client) GetMetricIngestor(ctx context.Context, tableName string) (ingest.Ingestor, error) {
+	if ingestor := adx.ingestors[tableName]; ingestor != nil {
+		return ingestor, nil
+	}
+
+	if adx.cfg.CreateTables {
+		if _, err := adx.client.Mgmt(ctx, adx.cfg.Database, createTableCommand(tableName)); err != nil {
+			return nil, fmt.Errorf("creating table for %q failed: %w", tableName, err)
+		}
+
+		if _, err := adx.client.Mgmt(ctx, adx.cfg.Database, createTableMappingCommand(tableName)); err != nil {
+			return nil, err
+		}
+	}
+
+	// Create a new ingestor client for the table
+	var ingestor ingest.Ingestor
+	var err error
+	switch strings.ToLower(adx.cfg.IngestionType) {
+	case ManagedIngestion:
+		ingestor, err = ingest.NewManaged(adx.client, adx.cfg.Database, tableName)
+	case QueuedIngestion:
+		ingestor, err = ingest.New(adx.client, adx.cfg.Database, tableName, ingest.WithStaticBuffer(bufferSize, maxBuffers))
+	default:
+		ingestor, err = nil, fmt.Errorf(`ingestion_type has to be one of %q or %q`, ManagedIngestion, QueuedIngestion)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("creating ingestor for %q failed: %w", tableName, err)
+	}
+	adx.ingestors[tableName] = ingestor
+
+	return ingestor, nil
+}
+
+func createTableCommand(table string) kusto.Statement {
+	builder := kql.New(`.create-merge table ['`).AddTable(table).AddLiteral(`'] `)
+	builder.AddLiteral(`(['fields']:dynamic, ['name']:string, ['tags']:dynamic, ['timestamp']:datetime);`)
+
+	return builder
+}
+
+func createTableMappingCommand(table string) kusto.Statement {
+	builder := kql.New(`.create-or-alter table ['`).AddTable(table).AddLiteral(`'] `)
+	builder.AddLiteral(`ingestion json mapping '`).AddTable(table + "_mapping").AddLiteral(`' `)
+	builder.AddLiteral(`'[{"column":"fields", `)
+	builder.AddLiteral(`"Properties":{"Path":"$[\'fields\']"}},{"column":"name", `)
+	builder.AddLiteral(`"Properties":{"Path":"$[\'name\']"}},{"column":"tags", `)
+	builder.AddLiteral(`"Properties":{"Path":"$[\'tags\']"}},{"column":"timestamp", `)
+	builder.AddLiteral(`"Properties":{"Path":"$[\'timestamp\']"}}]'`)
+
+	return builder
+}
+
+// Setters for testing
+func (adx *Client) SetLogger(logger telegraf.Logger) {
+	adx.logger = logger
+}
+
+func (adx *Client) SetKustoClient(client *kusto.Client) {
+	adx.client = client
+}
+
+func (adx *Client) SetMetricsIngestors(ingestors map[string]ingest.Ingestor) {
+	adx.ingestors = ingestors
+}
+
+func (adx *Client) SetConfig(config *Config) {
+	adx.cfg = config
+}
```

Comment:
Please remove those! The internal tests are in the same package and therefore you can easily set internal fields. For the plugin tests, the only thing that needs to be "faked" is the `client` so how about adding a
```golang
MockClient bool `toml:"-"`
```
to the `Config` which can be used to make `NewClient` to set `client` to the mock version. This has the advantage that the field cannot be set by the user but only "internally" and you can use the usual flow in the plugin without needing to do "special things" in the plugin.